Packages can constrain k8s and kc versions

joe-kimmel-vmw opened the pull request:
- kubernetes and kapp-controller version constraints can be overridden by annotations
- range version selection of a Package chooses the highest version of
  that package which also satisfies kc and k8s version constraints
- Error messages if no package satisfies constraints provides detail of
  which constraints failed.
- if PKGI specifies a different cluster, that cluster's version of k8s
  is checked for the constraints (manually verified - no automated test
  of this)

<!--  Thanks for sending a pull request!  Here are some tips for you:

If this is your first time, please read our contributor guidelines: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/CONTRIBUTING.md and developer guide https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/docs/dev.md
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
If no issue exists for this change, please create an issue and link it here.
-->
Fixes #484 
Fixes #577 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. 

-->
```release-note
Packages can constrain kubernetes and kapp-controller versions
```

#### Additional Notes for your reviewer:

note that we def. punted on the multicluster case test.  We did verify this manually several times throughout development and again after our last change.  However, open to discussion re. using some shell scripts or introducing an e2e dependency on CAPI for spinning up a second cluster and making that cluster's kubeconfig available to kc via a Secret.

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

design proposal doc: https://hackmd.io/6FQhzRIPR_uceqtTqSIXaA

carvel.dev docs pending

```docs

```

joe-kimmel-vmw commented:
note that @neil-hickey coauthored this PR with me.

cppforlife commented on config/crds.yml:
```diff
@@ -108,6 +108,32 @@ spec:
+                properties:
+                  constraints:
+                    type: string
+                  prereleases:
+                    properties:
+                      identifiers:
+                        items:
+                          type: string
+                        type: array
+                    type: object
```
lets collect various versions of k8s distros so that we know what is the versions 'data set' we are working with. on gke all versions are prereleases and so it would be somewhat awkward to ask all users to set `prereleases: {}`

joe-kimmel-vmw replied:
azure looks like they just use major.minor.patch: https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli

joe-kimmel-vmw replied:
EKS also looks like it uses major.minor.patch: https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html

neil-hickey replied:
- TKGm : (don't seem to use a pre-release, but do include metadata about the build version)
```
NAME   STATUS   ROLES    AGE   VERSION
x      Ready    <none>    8d   v1.22.5+vmware.1
```
- AKS (azure): 
```
NAME   STATUS   ROLES    AGE   VERSION
x      Ready    <none>    8d   v1.23.8
```
- GKE (google):
```
NAME   STATUS   ROLES    AGE   VERSION
x      Ready    <none>    8d   v1.21.12-gke.2200
```
- Minikube:
```
NAME   STATUS   ROLES    AGE   VERSION
x      Ready    <none>    8d   v1.24.1
```
- EKS (amazon):
```
NAME  STATUS   ROLES    AGE   VERSION
x     Ready    <none>   10m   v1.21.12-eks-5308cf7
```
- OpenShift on vSphere:
- `v1.23.3+e419edf`

cppforlife commented on hack/dev-deploy.sh:
```diff
@@ -2,9 +2,13 @@
 
+source $(dirname "$0")/version-util.sh
+
+#ytt -f config/ -f config-release -v kapp_controller_version=" --data-values-env=KCTRL | kbld -f- > ./tmp/release.yml
```
remove

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,98 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
 	return reconcile.Result{}, nil
 }
 
+func (pi *PackageInstallCR) getClusterVersion() (*version.Info, error) {
```
let's move this back into deploy factory (as a new method: KubernetesEnv?) so that private details of it are not exposed to PackageInstall.

joe-kimmel-vmw replied:
fwiw we put it into deploy module but it still feels awkward to claim that it's part of a "factory" imo, - I already apologized to Neil for not just doing exactly what you said and causing more discussion, so - sorry, but i see your point re. scope of concern, but not re. "factory" specifically.

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,98 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
+	const kubernetesVersionOverrideAnnotation = "packaging.carvel.dev/ignore-kubernetes-version-selection"
+	_, found := pi.model.Annotations[kubernetesVersionOverrideAnnotation]
+	if found {
+		pi.log.Info("Found k8s Version override annotation; not applying version constraints")
```
nit: Version => version

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,98 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
+
+		if matchedVers.Len() == 0 {
+			return fmt.Errorf("Cluster is running kubernetes %s but package constrained versions to: %s",
+				vi.GitVersion, pkg.Spec.KubernetesVersionSelection.Constraints)
```
should probably use Description() -- https://github.com/vmware-tanzu/carvel-vendir/blob/develop/pkg/vendir/versions/v1alpha1/config.go#L39

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -241,7 +346,10 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
+			"Pre-Filter: Number of Packages excluded due to k8s or kc constraints: %d. (details: %v)\nPost-Exclusion Package Version Selection: %v",
+			len(rejectPkgReasons), rejectPkgReasons, err.Error())
+		return datapkgingv1alpha1.Package{}, retErr
 	}
```
ideally error would be something like this: `Expected to find at least one version, but did not (details: all=0 -> after-prereleases-filter=0 -> after-constraints-filter=0 -> after-k8s-compat-filter=0 -> after-kapp-controller-compat-filter=0)`

joe-kimmel-vmw replied:
pending https://github.com/vmware-tanzu/carvel-vendir/pull/175

cppforlife commented on pkg/packageinstall/reconciler.go:
```diff
@@ -22,6 +22,9 @@ import (
 )
 
+// ControllerVersion holds the semver string for the version of kapp-controller that is running, so that we can compare it against version constraints in packages
+var ControllerVersion string
```
lets get rid of a global variable in favor of injecting data into necessary class.

cppforlife commented on test/e2e/kappcontroller/packageinstall_test.go:
```diff
@@ -172,6 +173,234 @@ stringData:
+	return installPkgYAML
+}
+
+func Test_PackageInstalled_FromPackageInstall_VersionConstraints(t *testing.T) {
```
given the scope of the selection logic, it makes sense to push most of the variations into unit tests (PackageInstallCR class) and probably just keep one high level e2e test.

neil-hickey commented on go.mod:
```diff
@@ -4,13 +4,12 @@ go 1.18
 	github.com/prometheus/client_golang v1.11.1
 	github.com/stretchr/testify v1.7.5
-	github.com/vmware-tanzu/carvel-vendir v0.24.0
+	github.com/vmware-tanzu/carvel-vendir v0.29.1-0.20220727190601-9dd41a08ed60
```
This fairly large fast forward causes us to pull in more vendored things, hence why there are so many file changes...

neil-hickey commented on pkg/packageinstall/packageinstall_version_handler.go:
```diff
@@ -105,6 +105,7 @@ func (ipvh *PackageInstallVersionHandler) isEligibleForVersionUpgrade(version st
 
 	semverConfig := installedPkg.Spec.PackageRef.VersionSelection
 
+	// TODO: we should  make this apply the same kc and k8s constraints as the other one to avoid unnecessary reconciles
```
So this watcher doesn't get the PKGI resource,  only the PKG. This means:

- We can't check for the absense / presense of the ignore annotations
- We also have to duplicate / make public the constraint checking functions
- It introduces more complexity into this version handler and another API call / cross-reference check

Will probably want to chat this one through as a group

joe-kimmel-vmw commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -216,6 +289,9 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
 	if err != nil {
 		return datapkgingv1alpha1.Package{}, err
 	}
+	if len(pkgList.Items) == 0 {
```
we could add a test for this case.

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -52,6 +55,66 @@ func (s *ServiceAccounts) Find(genericOpts GenericOpts, saName string) (Processe
 	return pgoForSA, nil
 }
 
+// GetClusterVersion returns the kubernetes API version for the cluster which has been supplied to kapp-controller via a kubeconfig
```
lets move this code out to dedicated file cluster_version.go

cppforlife commented on pkg/deploy/service_accounts.go:
```diff
@@ -52,6 +55,66 @@ func (s *ServiceAccounts) Find(genericOpts GenericOpts, saName string) (Processe
+		saName,
+		specCluster,
+		GenericOpts{Name: objMeta.Name, Namespace: objMeta.Namespace},
+		NewServiceAccounts(cc, log),
+		NewKubeconfigSecrets(cc))
```
service accounts object should be shared with the one in the factory to take advantage of token cache.

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -26,6 +27,8 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
```
nit: remove newline

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,74 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
+		pi.log.Error(err, "Unable to retrieve cluster kubernetes version")
+		return false
+	}
+	v, err := dropPreAndBuild(vi.GitVersion)
```
dropPreAndBuild => dropVersionPrereleaseAndBuild might be a clearer name

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,74 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
+		pi.log.Error(err, "Unable to parse cluster kubernetes version", "version string", vi.GitVersion)
+		return false
+	}
+	constraints, err := semver.ParseRange(pkg.Spec.KubernetesVersionSelection.Constraints)
```
lets add a validation for constraints on a package: https://github.com/vmware-tanzu/carvel-kapp-controller/blob/e5667d3102f34e05df58e42670689b3d6ab5ee29/pkg/apiserver/apis/datapackaging/validation/validations.go#L37. with that addition, it's fine that we would ignore the error here (and log).

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -201,6 +205,74 @@ func (pi *PackageInstallCR) reconcileAppWithPackage(existingApp *kcv1alpha1.App,
+		return true
+	}
+
+	vi, err := deploy.GetClusterVersion(pi.coreClient, pi.model.Spec.ServiceAccountName, pi.model.Spec.Cluster, &pi.model.ObjectMeta, pi.log)
```
clusterVersionConstraintsSatisfied gets called per every package, so we should move out GetClusterVersion outside to avoid N calls.

cppforlife commented on test/e2e/kappcontroller/packageinstall_test.go:
```diff
@@ -172,6 +173,165 @@ stringData:
+				AllowError:  true})
+		assert.Contains(t, out, "after-kubernetes-version-check=0")
+		require.Error(t, err)
+		// Without calling cleanup after each stage, somehow it seems like the controller doesn't
+		// get the updated annotation.
+		cleanUp()
```
lets try to figure out since sounds like real users will run into this situation => update to annotation should "fix" the PackageInstall

cppforlife replied:
=> leave comment about metadata.generation not being incr when annotations change

joe-kimmel-vmw replied:
@cppforlife can we go over this one more time - still a little unclear to us exactly what is breaking and why it is/isn't a user-facing concern?

neil-hickey committed:
Packages can constrain k8s and kc versions

- kubernetes and kapp-controller version constraints can be overridden by annotations
- range version selection of a Package chooses the highest version of
  that package which also satisfies kc and k8s version constraints
- Error messages if no package satisfies constraints provides detail of
  which constraints failed.
- simplified error message for the case where you just have zero
  packages.
- kapp-controller version is threaded through into the PKGI reconciler
- refactor factories out to main thread, pass them down via dep injection (fixes a bug in the app factory where service account
  token cache was reinitialized each reconcile
- if PKGI specifies a different cluster, that cluster's version of k8s
  is checked for the constraints (manually verified - no automated test
  of this)
- Error on no packages found

Signed-off-by: Neil Hickey <nhickey@vmware.com>

joe-kimmel-vmw commented on cmd/controller/run.go:
```diff
@@ -165,6 +166,9 @@ func Run(opts Options, runLog logr.Logger) error {
 	refTracker := reftracker.NewAppRefTracker()
 	updateStatusTracker := reftracker.NewAppUpdateStatus()
 
+	// initialize deploy factory once - the deploy factory contains a cache which should be only setup once.
```
maybe worth saying "cache of _"  (service accounts? right?) to assist future readers in their journey

neil-hickey replied:
Yeh, we can make it more specific.

neil-hickey commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -239,7 +295,31 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
+		return pi.kcVersionConstraintsSatisfied(&pkg)
+	}
+
+	vi, err := pi.deployFactory.GetClusterVersion(pi.model.Spec.ServiceAccountName, pi.model.Spec.Cluster, &pi.model.ObjectMeta, pi.log)
```
Do we make the tradeoff of a two-pass filter vs a call to the cluster for k8s info on each reconcile....

By two pass I mean: first pass --> check if any packageinstall has constraints (if not, skip over this code), second pass as is currently.

cppforlife replied:
or it can also be memoized so that we get the best of both worlds.

neil-hickey replied:
Maybe it can be in the case of the on-cluster version. But should we really "cache" the version for an external cluster? Wouldn't that be prone to stale-ness?

cppforlife commented on pkg/app/app_factory.go:
```diff
@@ -29,13 +29,12 @@ type CRDAppFactory struct {
 
 // NewCRDApp creates a CRDApp injecting necessary dependencies.
-func (f *CRDAppFactory) NewCRDApp(app *kcv1alpha1.App, log logr.Logger) *CRDApp {
+func (f *CRDAppFactory) NewCRDApp(app *kcv1alpha1.App, log logr.Logger, deployFactory deploy.Factory) *CRDApp {
```
instead of passing down deployFactory into NewCRDApp, let's make CRDAppFactory constructor accept it. this way whoever calls NewCRDApp does not need to know anything about deployFactory.

neil-hickey replied:
By constructor do you mean just add it as another public field on the struct? There is no constructor right now for CRDAppFactory. We could make one, and then keep it as a private field?

neil-hickey replied:
@cppforlife

cppforlife commented on pkg/app/reconciler.go:
```diff
@@ -27,14 +28,21 @@ type Reconciler struct {
 }
 
 // NewReconciler constructs new Reconciler.
-func NewReconciler(appClient kcclient.Interface, log logr.Logger, crdAppFactory CRDAppFactory,
```
with change to CRDAppFactory accepting deployFactory, this class would not have to change (since deployFactory is an implementation detail of how CRDAppFactory works).

cppforlife commented on pkg/deploy/factory.go:
```diff
@@ -38,32 +41,42 @@ func NewFactory(coreClient kubernetes.Interface,
+
+// processOpts takes generic opts and a ServiceAccount Name, and returns a populated kubeconfig that can connect to a cluster.
+// if the saName is empty then you'll connect to a cluster via the clusterOpts inside the genericOpts, otherwise you'll use the specified SA.
+func processOpts(saName string, clusterOpts *v1alpha1.AppCluster, genericOpts GenericOpts, serviceAccounts *ServiceAccounts, kubeconfigSecrets *KubeconfigSecrets) (*ProcessedGenericOpts, error) {
```
since we have placed GetClusterVersion onto Factory, we should be able to reduce processOpts scope as well but placing it onto Factory. by doing that it will also not have to accept serviceAccounts and kubeconfigSecrets and just read off f.

cppforlife commented on pkg/deploy/factory.go:
```diff
@@ -38,12 +41,24 @@ func NewFactory(coreClient kubernetes.Interface,
+
+// processOpts takes generic opts and a ServiceAccount Name, and returns a populated kubeconfig that can connect to a cluster.
+// if the saName is empty then you'll connect to a cluster via the clusterOpts inside the genericOpts, otherwise you'll use the specified SA.
+func (f Factory) processOpts(saName string, clusterOpts *v1alpha1.AppCluster, genericOpts GenericOpts) (*ProcessedGenericOpts, error) {
```
nit: doesnt have to be pointer afaik

cppforlife commented on pkg/deploy/factory.go:
```diff
@@ -92,3 +105,33 @@ func (f Factory) globalKappDeployRawOpts() []string {
+// GetClusterVersion returns the kubernetes API version for the cluster which has been supplied to kapp-controller via a kubeconfig
+func (f Factory) GetClusterVersion(saName string, specCluster *v1alpha1.AppCluster, objMeta *metav1.ObjectMeta, log logr.Logger) (*version.Info, error) {
+	// this logic is duplicated here and in ProcessOpts: if the serviceAccount name is present, we will be deploying to this cluster
+	if len(saName) > 0 {
```
we should use a switch statement as above to catch the case when it's neither saName or kubeconfig

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -216,15 +273,22 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
 			versionStrs = append(versionStrs, pkg.Spec.Version)
 			versionToPkg[pkg.Spec.Version] = pkg
 		}
+		if pkgHasK8sConstraint(&pkg) {
+			requiresClusterVersion = true
+		}
```
this if should be within above if since we only care about a subset of packages.

cppforlife commented on pkg/packageinstall/packageinstall.go:
```diff
@@ -237,9 +301,35 @@ func (pi *PackageInstallCR) referencedPkgVersion() (datapkgingv1alpha1.Package,
 
-	selectedVersion, err := versions.HighestConstrainedVersion(versionStrs, verConfig)
+	vcc := []versions.ConstraintCallback{
+		{Constraint: k8sConstraint, Name: "kubernetes-version-check"},
```
should this constraint be only conditionally added when requiresClusterVersion is true? i dunno the impact of running clusterVersionConstraintsSatisfied with a blank version.

neil-hickey replied:
I guess the only consequence is that we will always have a constraint showing in the error message for the k8s version check, even if the user never specified that constraint,

neil-hickey replied:
I added a test to show this Test_Package_ConstraintNotGiven_ErrorDoesNotContainMessage
